Stackoverflow Exception

Hi,
I encountered a Stackoverflow exception in the scope manager. Can't really reproduce it since it appears to occur more or less random to me.
```
Caused by: java.lang.StackOverflowError
        at java.util.Objects.equals(Objects.java:77) ~[?:?]
        at de.fraunhofer.aisec.cpg.graph.types.ObjectType.equals(ObjectType.java:181) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs$lambda-15(ScopeManager.kt:584) ~[cpg-core-master-SNAPSHOT.jar:?]
        at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1631) ~[?:?]
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127) ~[?:?]
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230) ~[?:?]
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.noneMatch(ReferencePipeline.java:538) ~[?:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:582) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:581) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:581) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:581) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:581) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.passes.scopes.ScopeManager.getCurrentTypedefs(ScopeManager.kt:100) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.TypeManager.resolvePossibleTypedef(TypeManager.java:741) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.types.TypeParser.createFromUnsafe(TypeParser.java:846) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.types.TypeParser.createFrom(TypeParser.java:887) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.TypeManager.lambda$getCommonType$10(TypeManager.java:492) ~[cpg-core-master-SNAPSHOT.jar:?]
        at java.util.Optional.map(Optional.java:265) ~[?:?]
        at de.fraunhofer.aisec.cpg.graph.TypeManager.getCommonType(TypeManager.java:492) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.TypeManager.isSupertypeOf(TypeManager.java:570) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.lambda$setType$1(Expression.java:147) ~[cpg-core-master-SNAPSHOT.jar:?]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176) ~[?:?]
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.setType(Expression.java:148) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.BinaryOperator.typeChanged(BinaryOperator.java:200) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.lambda$setType$4(Expression.java:160) ~[cpg-core-master-SNAPSHOT.jar:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[?:?]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[?:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.setType(Expression.java:160) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.DeclaredReferenceExpression.typeChanged(DeclaredReferenceExpression.java:145) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.lambda$setType$4(Expression.java:160) ~[cpg-core-master-SNAPSHOT.jar:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[?:?]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[?:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.setType(Expression.java:160) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.BinaryOperator.typeChanged(BinaryOperator.java:200) ~[cpg-core-master-SNAPSHOT.jar:?]
        at de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression.lambda$setType$4(Expression.java:160) ~[cpg-core-master-SNAPSHOT.jar:?]
```
I guess thats a problem in the TypeManager or ScopeManager right?
I triggered that exception when parsing in the entire FFmpeg project and using UseParallelFrontend
Regards,
Tom

--- a/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/scopes/ScopeManager.kt
+++ b/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/scopes/ScopeManager.kt
@@ -96,7 +96,7 @@ class ScopeManager {
     val currentRecord: RecordDeclaration?
         get() = this.firstScopeIsInstanceOrNull<RecordScope>()?.astNode as? RecordDeclaration
 
-    val currentTypedefs: List<TypedefDeclaration>
+    val currentTypedefs: Collection<TypedefDeclaration>
         get() = this.getCurrentTypedefs(currentScope)
 
     val currentNamePrefix: String
@@ -572,23 +572,20 @@ class ScopeManager {
         }
     }
 
-    private fun getCurrentTypedefs(scope: Scope?): List<TypedefDeclaration> {
-        val curr: MutableList<TypedefDeclaration> = ArrayList()
-        if (scope is ValueDeclarationScope) {
-            curr.addAll(scope.typedefs)
-        }
-        if (scope!!.getParent() != null) {
-            for (parentTypedef in getCurrentTypedefs(scope.getParent())) {
-                if (curr.stream().map { obj: TypedefDeclaration -> obj.alias }.noneMatch { o: Type?
-                        ->
-                        parentTypedef.alias.equals(o)
-                    }
-                ) {
-                    curr.add(parentTypedef)
-                }
+    private fun getCurrentTypedefs(scope: Scope?): Collection<TypedefDeclaration> {
+        val typedefs = mutableMapOf<Type, TypedefDeclaration>()
+
+        var current = scope
+
+        while (current != null) {
+            if (scope is ValueDeclarationScope) {
+                scope.typedefs.forEach { typedefs.putIfAbsent(it.alias, it) }
             }
+
+            current = current.parent
         }
-        return curr
+
+        return typedefs.values
     }
 
     /**

